keps/sig-node/20190726-arbitrary-pull-secrets.md

```diff
@@ -0,0 +1,98 @@
+---
+title: Arbitrary image pull secrets
+authors:
+  - "@sashayakovtseva"
+owning-sig: sig-node
+creation-date: 2019-07-26
+status: provisional
+---
+
+# Arbitrary image pull secrets
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+  - [Implementation Details](#implementation-details)
+- [Alternatives ](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Currently in order to pull an image from a private registry one can use `imagePullSecrets` (as described
+[here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/)). 
+However, it is tightly integrated with Docker and there is no other way to support different on-prem
+registries or different types of authorization (e.g. private tokens).
+
+This proposal suggests changes that will fully enable arbitrary private image registry support.
+
+## Motivation
+
+Kubernetes introduced CRI making it possible to integrate any container runtime with it.
+CRI contains `ImageService` that is responsible for any image-related operations performed by kubelet. 
+
+Whenever a new container is created and base image is not present on node, `PullImage` method of `ImageService`
+is called. `PullImageRequest` contains `AuthConfig`, so no changes should be make in CRI to enable support
+for private image pulls.   
+
+It is common to have private container images or on-prem image registries when working with Kubernetes. To use
+those, one should either perform node configuration or leverage `imagePullSecrets` for pod that will be passed
+to the CRI implementation in `AuthConfig`.
+ 
+However, even though Kubernetes has CRI, `imagePullSecrets` are respected only when the secret type is
+`kubernetes.io/dockerconfigjson` or `kubernetes.io/dockercfg`. This is not aligned with CRI concept since CRI should
+enable support for arbitrary container image registries (via `ImageService`) and arbitrary container runtimes (via
+`RuntimeService`). 
+
+### Goals
+
+- Introduce a new secret type, e.g. `kubernetes.io/regcredentials`.
+It should contain data that can be mapped to the CRI's `AuthConfig` subset of fields directly.
+  
+- Do not interpret passed credentials as Docker’s when a secret of type `kubernetes.io/regcredentials`
+is used in `imagePullSecrets`. Pass it directly to the CRI implementation and let it decide how to handle them instead.
```

Comment:
I don’t see a reason to make the secret opaque.  We can always change the auth config spec or introduce a new secret type - opaque secrets are against our general approach to defining points of consistency between container runtimes.

I would be supportive of letting container runtimes interpret other information about a pod in order to decide on secret values, but we’d need to be more concrete in this proposal about what that actually means.